zipline/sources/data_frame_source.py

```diff
@@ -108,11 +119,21 @@ class DataPanelSource(DataSource):
     def __init__(self, data, **kwargs):
         assert isinstance(data.major_axis, pd.tseries.index.DatetimeIndex)
 
-        self.data = data
+        self.data = data.fillna(method='ffill', axis=0)
         # Unpack config dictionary with default values.
-        self.sids = kwargs.get('sids', data.items)
-        self.start = kwargs.get('start', data.major_axis[0])
-        self.end = kwargs.get('end', data.major_axis[-1])
+        self.start = kwargs.get('start', self.data.major_axis[0])
+        self.end = kwargs.get('end', self.data.major_axis[-1])
+
+        # Remap sids based on the trading environment
+        self.identifiers = kwargs.get('sids', self.data.items)
+        trading.environment.update_asset_finder(identifiers=self.identifiers)
+        self.data.items = [
+            trading.environment.asset_finder.lookup_generic(
+                identifier,
+                as_of_date=self.end)[0].sid
+            for identifier in self.data.items
+        ]
+        self.sids = self.data.items
```

Comment:
@jfkirk I don't fully understand the question being asked here.  Can you come grab me at some point this afternoon and we'll talk about this?